src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -206,11 +206,14 @@ void BackgroundNotificationCb (NSNotification obj)
 			// iteration. We split the operation in two, get all the diff cancelation sources, then try to cancel each of them
 			// which will do the correct lock dance. Note that we could be tempted to do a RemoveAll, that will yield the same
 			// runtime issue, this is dull but safe. 
-			var sources = new List <TaskCompletionSource<HttpResponseMessage>> (inflightRequests.Count);
-			foreach (var r in inflightRequests.Values) {
-				sources.Add (r.CompletionSource);
+			List <TaskCompletionSource<HttpResponseMessage>> sources = null; 
+			lock (inflightRequestsLock) { // just lock when we iterate
+				sources = new List <TaskCompletionSource<HttpResponseMessage>> (inflightRequests.Count);
+				foreach (var r in inflightRequests.Values) {
+					sources.Add (r.CompletionSource);
+				}
 			}
-			sources.ForEach (source => { source.TrySetCanceled (); });
+			sources?.ForEach (source => { source.TrySetCanceled (); });
```

Comment:
`sources` should never be `null` here

Reply:
Agreed, but I have 0 trust ;)

Reply:
Just like me :) but adding `?` means you're hiding a problem (if any) which will only make diagnosing it much harder than it needs to be.

IOW my lack of trust makes me rather have this _should-never-be-null_ value throw an NRE. I rather have a bug I understand than one I don't ;-)